File: enode/src/main/java/com/enode/domain/impl/DefaultAggregateRootInternalHandlerProvider.java
package com.enode.domain.impl;

import com.enode.common.function.Action2;
import com.enode.domain.IAggregateRoot;
import com.enode.domain.IAggregateRootInternalHandlerProvider;
import com.enode.eventing.IDomainEvent;
import com.enode.infrastructure.IAssemblyInitializer;
import com.enode.infrastructure.TypeUtils;
import com.enode.infrastructure.WrappedRuntimeException;

import java.lang.invoke.MethodHandle;
import java.lang.invoke.MethodHandles;
import java.lang.reflect.Method;
import java.lang.reflect.Modifier;
import java.util.Arrays;
import java.util.HashMap;
import java.util.Map;
import java.util.Set;

import static com.enode.common.Constants.AGGREGATE_ROOT_HANDLE_METHOD_NAME;

public class DefaultAggregateRootInternalHandlerProvider implements IAggregateRootInternalHandlerProvider, IAssemblyInitializer {

    private Map<Class, Map<Class, Action2<IAggregateRoot, IDomainEvent>>> mappings = new HashMap<>();

    @Override
    public void initialize(Set<Class<?>> componentTypes) {
        componentTypes.stream().filter(TypeUtils::isAggregateRoot).forEach(this::recurseRegisterInternalHandler);
    }

    private void recurseRegisterInternalHandler(Class aggregateRootType) {
        Class superclass = aggregateRootType.getSuperclass();

        if (!isInterfaceOrObjectClass(superclass)) {
            registerInternalHandlerWithSuperclass(aggregateRootType, superclass);
        }

        register(aggregateRootType, aggregateRootType);
    }

    private void registerInternalHandlerWithSuperclass(Class aggregateRootType, Class parentType) {
        Class superclass = parentType.getSuperclass();

        if (!isInterfaceOrObjectClass(superclass)) {
            registerInternalHandlerWithSuperclass(aggregateRootType, superclass);
        }

        register(aggregateRootType, parentType);

    }

    private boolean isInterfaceOrObjectClass(Class type) {
        return Modifier.isInterface(type.getModifiers()) || type.equals(Object.class);
    }

    private void register(Class aggregateRootType, Class type) {
        Arrays.asList(type.getDeclaredMethods()).stream()
                .filter(method ->
                        method.getName().equalsIgnoreCase(AGGREGATE_ROOT_HANDLE_METHOD_NAME)
                                && method.getParameterTypes().length == 1
                                && IDomainEvent.class.isAssignableFrom(method.getParameterTypes()[0])
                )
                .forEach(method -> registerInternalHandler(aggregateRootType, method.getParameterTypes()[0], method)
                );
    }

    private void registerInternalHandler(Class aggregateRootType, Class eventType, Method method) {
        Map<Class, Action2<IAggregateRoot, IDomainEvent>> eventHandlerDic = mappings.computeIfAbsent(aggregateRootType, k -> new HashMap<>());
        try {
            method.setAccessible(true);
            MethodHandle methodHandle = MethodHandles.lookup().unreflect(method);
            eventHandlerDic.put(eventType, (aggregateRoot, domainEvent) -> {
                try {
                    methodHandle.invoke(aggregateRoot, domainEvent);
                } catch (Throwable throwable) {
                    throw new RuntimeException(throwable);
                }
            });
        } catch (Exception e) {
            throw new WrappedRuntimeException(e);
        }
    }

    @Override
    public Action2<IAggregateRoot, IDomainEvent> getInternalEventHandler(Class<? extends IAggregateRoot> aggregateRootType, Class<? extends IDomainEvent> anEventType) {
        Class currentAggregateType = aggregateRootType;
        while (currentAggregateType != null) {
            Action2<IAggregateRoot, IDomainEvent> handler = getEventHandler(currentAggregateType, anEventType);
            if (handler != null) {
                return handler;
            }
            if (currentAggregateType.getSuperclass() != null && Arrays.asList(currentAggregateType.getSuperclass().getInterfaces()).contains(IAggregateRoot.class)) {
                currentAggregateType = currentAggregateType.getSuperclass();
            } else {
                break;
            }
        }
        return null;
    }

    private Action2<IAggregateRoot, IDomainEvent> getEventHandler(Class<? extends IAggregateRoot> aggregateRootType, Class<? extends IDomainEvent> anEventType) {
        Map<Class, Action2<IAggregateRoot, IDomainEvent>> eventHandlerDic = mappings.get(aggregateRootType);
        if (eventHandlerDic == null) {
            return null;
        }
        return eventHandlerDic.get(anEventType);
    }
}

File: enode/src/main/java/org/enodeframework/domain/impl/DefaultAggregateRootInternalHandlerProvider.kt
package org.enodeframework.domain.impl

import org.enodeframework.common.exception.HandlerNotFoundException
import org.enodeframework.common.function.Action2
import org.enodeframework.configurations.SysProperties
import org.enodeframework.domain.AggregateRoot
import org.enodeframework.domain.AggregateRootInternalHandlerProvider
import org.enodeframework.eventing.DomainEventMessage
import org.enodeframework.infrastructure.AssemblyInitializer
import org.enodeframework.infrastructure.TypeUtils
import java.lang.invoke.MethodHandles
import java.lang.reflect.Method
import java.lang.reflect.Modifier
import java.util.*

/**
 * @author anruence@gmail.com
 */
class DefaultAggregateRootInternalHandlerProvider : AggregateRootInternalHandlerProvider, AssemblyInitializer {

    private val aggregateRootInternalHandlerMap: MutableMap<Class<*>, MutableMap<Class<*>, Action2<AggregateRoot, DomainEventMessage<*>>>> =
        HashMap()

    override fun initialize(componentTypes: Set<Class<*>>) {
        componentTypes.filter { type: Class<*> -> TypeUtils.isAggregateRoot(type) }
            .forEach { aggregateRootType: Class<*> -> recurseRegisterInternalHandler(aggregateRootType) }
    }

    private fun recurseRegisterInternalHandler(aggregateRootType: Class<*>) {
        val superclass = aggregateRootType.superclass
        if (!isInterfaceOrObjectClass(superclass)) {
            registerInternalHandlerWithSuperclass(aggregateRootType, superclass)
        }
        register(aggregateRootType, aggregateRootType)
    }

    private fun registerInternalHandlerWithSuperclass(aggregateRootType: Class<*>, parentType: Class<*>) {
        val superclass = parentType.superclass
        if (!isInterfaceOrObjectClass(superclass)) {
            registerInternalHandlerWithSuperclass(aggregateRootType, superclass)
        }
        register(aggregateRootType, parentType)
    }

    private fun isInterfaceOrObjectClass(type: Class<*>): Boolean {
        return Modifier.isInterface(type.modifiers) || type == Any::class.java
    }

    private fun register(aggregateRootType: Class<*>, type: Class<*>) {
        Arrays.stream(type.declaredMethods).filter { method: Method ->
            method.name.startsWith(SysProperties.AGGREGATE_ROOT_HANDLE_METHOD_NAME)
                    && method.parameterTypes.size == 1 && DomainEventMessage::class.java.isAssignableFrom(method.parameterTypes[0])
        }.forEach { method ->
            registerInternalHandler(aggregateRootType, method.parameterTypes[0], method)
        }
    }

    private fun registerInternalHandler(aggregateRootType: Class<*>, eventType: Class<*>, method: Method) {
        val eventHandlerDic = aggregateRootInternalHandlerMap.computeIfAbsent(aggregateRootType) { HashMap() }
        method.isAccessible = true
        val methodHandle = MethodHandles.lookup().unreflect(method)
        eventHandlerDic[eventType] =
            Action2 { aggregateRoot: AggregateRoot, domainEventMessage: DomainEventMessage<*> ->
                methodHandle.invoke(aggregateRoot, domainEventMessage)
            }
    }

    override fun getInternalEventHandler(
        aggregateRootType: Class<out AggregateRoot>,
        eventType: Class<out DomainEventMessage<*>>
    ): Action2<AggregateRoot, DomainEventMessage<*>> {
        var currentAggregateType = aggregateRootType
        while (true) {
            val handler = aggregateRootInternalHandlerMap[currentAggregateType]?.get(eventType)
            if (handler != null) {
                return handler
            }
            if (currentAggregateType.superclass != null
                && listOf(*currentAggregateType.superclass.interfaces).contains(AggregateRoot::class.java)
            ) {
                currentAggregateType = currentAggregateType.superclass as Class<out AggregateRoot>
            } else {
                break
            }
        }
        throw HandlerNotFoundException(
            String.format(
                "Could not find event handler for [%s] of [%s]",
                eventType.javaClass.name,
                aggregateRootType.javaClass.name
            )
        )
    }
}